can/io/mf4.py

```diff
@@ -0,0 +1,392 @@
+
+try:
+    from asammdf import MDF4, Signal
+    import numpy as np
+
+    ASAMMDF_AVAILABLE = True
+
+except ImportError as error:
+    ASAMMDF_AVAILABLE = False
+    raise error
+
+
+CAN_MSG_EXT = 0x80000000
+CAN_ID_MASK = 0x1FFFFFFF
+
+STD_DTYPE = np.dtype(
+    [
+        ("CAN_DataFrame.BusChannel", "<u1"),
+        ("CAN_DataFrame.ID", "<u4"),
+        ("CAN_DataFrame.IDE", "<u1"),
+        ("CAN_DataFrame.DLC", "<u1"),
+        ("CAN_DataFrame.DataLength", "<u1"),
+        ("CAN_DataFrame.DataBytes", "(64,)u1"),
+        ("CAN_DataFrame.Dir", "<u1"),
+        ("CAN_DataFrame.EDL", "<u1"),
+        ("CAN_DataFrame.BRS", "<u1"),
+        ("CAN_DataFrame.ESI", "<u1"),
+    ]
+)
+
+ERR_DTYPE = np.dtype(
+    [
+        ("CAN_ErrorFrame.BusChannel", "<u1"),
+        ("CAN_ErrorFrame.ID", "<u4"),
+        ("CAN_ErrorFrame.IDE", "<u1"),
+        ("CAN_ErrorFrame.DLC", "<u1"),
+        ("CAN_ErrorFrame.DataLength", "<u1"),
+        ("CAN_ErrorFrame.DataBytes", "(64,)u1"),
+        ("CAN_ErrorFrame.Dir", "<u1"),
+        ("CAN_ErrorFrame.EDL", "<u1"),
+        ("CAN_ErrorFrame.BRS", "<u1"),
+        ("CAN_ErrorFrame.ESI", "<u1"),
+    ]
+)
+
+RTR_DTYPE = np.dtype(
+    [
+        ("CAN_DataFrame.BusChannel", "<u1"),
+        ("CAN_DataFrame.ID", "<u4"),
+        ("CAN_DataFrame.IDE", "<u1"),
+        ("CAN_DataFrame.DLC", "<u1"),
+        ("CAN_DataFrame.Dir", "<u1"),
+    ]
+)
+
+FD_LEN2DLC = {12: 9, 16: 10, 20: 11, 24: 12, 32: 13, 48: 14, 64: 15}
+
+FD_DLC2LEN = {value: key for key, value in FD_LEN2DLC.items()}
+
+logger = logging.getLogger("can.io.mf4")
+
+
+class MF4Writer(BaseIOHandler, Listener):
+    """Logs CAN data to an ASAM Measurement Data File v4 (.mf4).
+
+    MF4Writer does not support append mode.
+
+    If a message has a timestamp smaller than the previous one or None,
+    it gets assigned the timestamp that was written for the last message.
+    It the first message does not have a timestamp, it is set to zero.
+    """
+
+    def __init__(self, file, database=None):
+        """
+        :param file: a path-like object or as file-like object to write to
+                        If this is a file-like object, is has to be opened in
+                        binary write mode, not text write mode.
+        :param database: optional path to a DBC or ARXML file that contains
+                            message description.
+        """
+        super().__init__(file, mode="r+b")
+        now = datetime.now()
+        self._mdf = MDF4()
+        self._mdf.header.start_time = now
+        self.last_timestamp = self._start_time = now.timestamp()
+
+        if database:
+            database = Path(database).resolve()
+            if database.exists():
+                data = database.read_bytes()
+                attachment = data, database.name
+            else:
+                attachment = None
+        else:
+            attachment = None
+
+        # standard frames group
+        self._mdf.append(
```

Comment:
Since AFAIK, MDF/M4F files don't have a specified file structure (ie. you can specify channels/channel groups however you like to suit your application), does it really make sense to provide a default vs. the user implementing their own?

Reply:
I thought the idea was top have this ready to use Writer that the users can use without knowing the details of the MDF format.